### PR TITLE
Allow Spring generated code to use new OAS 3 annotations

### LINKED ISSUE #9774: [REQ][Java][Spring] Allow Spring generated code to use new OAS 3 annotations
##### Description

The current generation of Spring generates annotations using `io.swagger.annotations.*`

New versions of Spring and Swagger can use `io.swagger.v3.oas.annotations.*`

##### openapi-generator version

5.2.0

##### Command line used for generation

`java -jar ./openapi-generator-cli.jar generate -g spring -o generated_new -i ./openapi-specifications/client-user-manager/clientusermanager_openapi.yaml -c ./openapi-spring-config.yaml`

```yaml
dateLibrary: java8
library: spring-boot
serializableModel: true
hideGenerationTimestamp: true
modelNameSuffix: DTO
useOneOfInterfaces: true
addOneOfInterfaceImports: true
useTags: true
basePackage: co.ritual.clientusermanager
apiPackage: co.ritual.clientusermanager.generated.api
invokerPackage: co.ritual.clientusermanager.generated.api
modelPackage: co.ritual.clientusermanager.generated.models
configPackage: co.ritual.clientusermanager.generated.configuration
```

##### Steps to reproduce

Run the generator

##### Related issues/PRs

https://github.com/OpenAPITools/openapi-generator/issues/9604 - Uses Springfox annotations

##### Suggest a fix/enhancement

Add in options to use OAS3 annotations / imports as well as allowing the API to be marked as a `@Controller`

Explicitly allow setting Springfox to be disabled

Proposed fix: https://github.com/OpenAPITools/openapi-generator/pull/9775

### PR DESCRIPTION
**Pull request by welshm**
- Add in `oas3` option for Spring codegen to use newer annotations
- Add `useSpringController` option for Spring codegen
- Use `useSpringfox` to fix some unwanted imports for Spring codegen
- Use `jdk8` to add OffsetDateTime import for models in Spring codegen
- Add `JsonValue` to `enumOuterClass.mustache` to allow enums to be
  generated properly

- To test, use `oas3: true` `useSpringController: true` and `useSpringfox: false` in a config file to generate a Spring target.

To close #9774

### PR checklist
 
- [x] Read the [contribution guidelines](https://github.com/openapitools/openapi-generator/blob/master/CONTRIBUTING.md).
- [x] Pull Request title clearly describes the work in the pull request and Pull Request description provides details about how to validate the work. Missing information here may result in delayed response from the community.
- [x] Run the following to [build the project](https://github.com/OpenAPITools/openapi-generator#14---build-projects) and update samples:
  ```
  ./mvnw clean package 
  ./bin/generate-samples.sh
  ./bin/utils/export_docs_generators.sh
  ``` 
- [x] File the PR against the [correct branch](https://github.com/OpenAPITools/openapi-generator/wiki/Git-Branches): `master`, `5.1.x`, `6.0.x`
- [x] If your PR is targeting a particular programming language, @mention the [technical committee](https://github.com/openapitools/openapi-generator/#62---openapi-generator-technical-committee) members, so they are more likely to review the pull request.

### CODE REVIEW

**Comment by welshm** [comment]
I think @diyfr and @nmuesch might be good reviewers for this?

File: modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/SpringCodegen.java [hunk 1]
```diff
@@ -452,8 +473,10 @@ public void processOpts() {
         }
 
 
-        if (!this.apiFirst && !this.reactive) {
-            additionalProperties.put("useSpringfox", true);
+        if (this.useSpringfox) {
+            if (!this.apiFirst && !this.reactive) {
```
Comment (ritual-nik) [inline modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/SpringCodegen.java:477]: minor, can simplify to: this.useSpringFox && !this.apiFirst && !this.reactive

Comment (welshm) [reply modules/openapi-generator/src/main/java/org/openapitools/codegen/languages/SpringCodegen.java:477]: I'll add comments - I assume that springFox cannot be used when apiFirst and reactive is enabled but we don't use those options so I wanted it to be clear that this check is separate from the `useSpringFox` option check

**Comment by welshm** [comment]
@daonomic or @lwlee2608 Would you be an appropriate reviewer for this?

**Comment by welshm** [comment]
@wing328 - wondering if you can help me find an appropriate reviewer or how I might go about getting this added?

**Comment by welshm** [comment]
@4brunu Would you be able to help me get a review started on this? I have not had responses from anyone else so far

**Comment by 4brunu** [comment]
You need to copy the java technical committee members for review.

@bbdouglas (2017/07) @sreeshas (2017/08) @jfiala (2017/08) @lukoyanov (2017/09) @cbornet (2017/09) @jeff9finger (2018/01) @karismann (2019/03) @Zomzog (2019/04) @lwlee2608 (2019/10) @nmuesch (2021/01)

**Comment by welshm** [comment]
I'm happy to join or participate in a technical committee if that might help...

**Comment by welshm** [comment]
@4brunu do you happen to know what the expected response time is for opening a PR and receiving a review or for applying for the technical commmittee?

Part of wanting to contribute back is so that my organization doesn't diverge too much from the project here, but it's now been over a month with no response to opening the PR other than from yourself...

**Comment by wing328** [comment]
Let me try to take a look tomorrow or later this week.

**Comment by welshm** [comment]
> Let me try to take a look tomorrow or later this week.

Just following up here - I will be unavailalbe Aug 3 - 6 if there is feedback on this PR.

I am very much interested in joining the technical committee for Spring and Java if you're taking applicants (I did send an email) in order to help keep the generator supported for those languages.

**Comment by welshm** [comment]
@wing328 any chance to get this reviewed?

**Comment by welshm** [comment]
@wing328 Updated with latest master and regenerated tests/samples

**Comment by MichaelKunze** [comment]
Would be cool if this PR gets considered! Need this too.

**Comment by welshm** [comment]
@wing328 @4brunu is there anything more I can do to get this reviewed or looked at by someone on the team that can approve this?

This PR has been open for 2+ months, and I've kept it current with the main branch several times, but if no one will review it then there is no value in me maintaining this branch/PR.

I am more than happy to get more involved in the project, but my emails asking to join the team have also gone unanswered.

**Comment by wing328** [comment]
@welshm first of all. Thanks for the PR. We definitely want to review and merge this PR but we're short-handed as you can see there's a backlog of PRs pending review.

There are a few spring-related PRs that I plan to review later this week. I'll let you know once I've reviewed those and would need your help to resolve the merge conflicts one more time before a final review early or later next week. Does it sound like a plan?

**Comment by welshm** [comment]
@wing328 Works for me - I'm happy to deal with merge conflicts whenever. If there are more Spring related changes coming then I can look at resolving the conflicts early next week.

Thanks for the speedy reply! In relation to the short-handed-ness, do let me know if there is anything I can do to support / help / join the team.

**Comment by wing328** [comment]
I've merged a few spring-related PRs into master.

Can you please merge the latest master into your branch and resolve the merge conflicts?

I'll then review this weekend. Thanks.

**Comment by welshm** [comment]
Updated to latest master and resolved merge conflicts

**Comment by welshm** [comment]
I think it was actually due to a bad merge from master that did not change `{{{description}}}` to `{{{.}}}` for pojo.mustache - which I've now updated and hopefully will pass

**Comment by wing328** [comment]
Looks like it's still an issue:
```
[ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/server/petstore/spring-mvc/src/main/java/org/openapitools/model/XmlItem.java:132: error: ')' expected
[ERROR] @ApiModelProperty((example = "string", value = "")
[ERROR] ^
[ERROR] /home/circleci/OpenAPITools/openapi-generator/samples/server/petstore/spring-mvc/src/main/java/org/openapitools/model/XmlItem.java:152: error: ')' expected
[ERROR] @ApiModelProperty((example = "1.234", value = "")
[ERROR] ^
[ERROR] 
[ERROR] Command line was: /usr/lib/jvm/java-8-openjdk-amd64/jre/../bin/javadoc @options @packages
[ERROR] 
[ERROR] Refer to the generated Javadoc files in '/home/circleci/OpenAPITools/openapi-generator/samples/server/petstore/spring-mvc/target/site/apidocs' dir.
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
[ERROR] 
[ERROR] After correcting the problems, you can resume the build with the command
[ERROR]   mvn <goals> -rf :spring-mvc-server
2021-09-02 12:54:39.088:INFO::main: Logging initialized @136ms
2021-09-02 12:54:39.109:INFO:oejr.Runner:main: Runner
2021-09-02 12:54:39.264:INFO:oejs.Server:main: jetty-9.2.9.v20150224
Sep 02, 2021 12:54:43 PM com.sun.jersey.api.core.PackagesResourceConfig init
```

Please take a look when you've time.

**Comment by welshm** [comment]
Yeah, saw that `pojo.mustache` had the extra `(` still - removed and hopefully this run passes

**Comment by welshm** [comment]
Just a heads up, I will be travelling and unavailable until Sept 7 to address any feedback or comments.

Thanks!

**Comment by wing328** [comment]
Enjoy your travel. I'll merge a few minor spring-related PRs while we perform more tests with your PR.

**Comment by welshm** [comment]
Back and more than happy and willing to make changes here as needed :)

**Comment by welshm** [comment]
Just wanted to follow up here - happy to rebase/merge again as needed

**Comment by atkawa7** [comment]
ping @wing328

**Comment by welshm** [comment]
@wing328 Up to date - will pay attention to CI results

**Comment by atkawa7** [comment]
@welshm Thank you. Cannot wait on this PR @wing328

**Comment by welshm** [comment]
Should be resolved, it was using additional params based on if the key was present at all instead of evaluating the actual value - fixed by using `writePropertyBack`

**Comment by welshm** [comment]
Looks like circleCI failed, but I don't see a test failure - looks like maybe a web thing?

**Comment by welshm** [comment]
Fixed a few more issues with the generator (leftover `@Parameter` needed to be converted from `@ApiParam` - and some allowableValue annotations needed to be switched to `@Scheme`)

I did _not_ fix the generated `pom` files, but I'm happy to tackle that if it's required. When I looked, it seems to automatically include `springfox` regardless of setting, which seems odd.

I'm personally using Bazel for my projects, so it will take me a bit more time to fix the maven if desired

**Comment by wing328** [comment]
Did a few more tests and the results are good so let's merge this one and file separate PRs for enhancements if needed.

**Comment by atkawa7** [comment]
@wing328 would you be able to do a minor release to test this out

**Comment by wing328** [comment]
@atkawa7 can you please use the SNAPSHOT version (links in the readme) before v5.3.1 release (which is scheduled a month later)?

**Comment by atkawa7** [comment]
Thanks @wing328

**Comment by mhinders** [comment]
Really happy about this change. I would however be inclined to use springdocs for viewing the API definition. As it is currently the Spring controllers are generated with @org.springframework.stereotype.Controller and not @org.springframework.web.bind.annotation.RestController which means that manual configuration is required as explained here https://springdoc.org/#my-rest-controller-using-controller-annotation-is-ignored. Could the templates be changed to use RestController?

**Comment by welshm** [comment]
> Really happy about this change. I would however be inclined to use springdocs for viewing the API definition. As it is currently the Spring controllers are generated with @org.springframework.stereotype.Controller and not @org.springframework.web.bind.annotation.RestController which means that manual configuration is required as explained here https://springdoc.org/#my-rest-controller-using-controller-annotation-is-ignored. Could the templates be changed to use RestController?

It can be updated to do that - just for clarity though, I had added this as an option for the interfaces but had left the generated API controller implementations (`apiController.mustache`) unchanged - would you be expecting both to be annotated with `@RestController`? My opinion is probably not, as the interface has all the annotations you need.

**Comment by BlackSunshine-manage** [comment]
Hello! how i can use this plugin?? this is a big work and i waited to long time this. You're really cool!

**Comment by Jey2k** [comment]
Hi,
Quite a noob in this part, but I'm currently trying to migrate one of our application and I would like to use this generator. So first thanks for your work.
The last problem detected is with "oas3" activated, and a default value on a query param it gives some code not compatible with Swagger v3 annotations:
eg: 
`,@Parameter(name = "size", description = "Size of the results page (maximum size is 1'000)", defaultValue = "1000") @RequestParam(value = "size", required = false, defaultValue = "1000") Integer size`

I suppose it should be included into @Schema, isn't it ?
eg:
`,@Parameter(name = "size", description = "Size of the results page (maximum size is 1'000)", schema =  @Schema( defaultValue = "1000" )) @RequestParam(value = "size", required = false, defaultValue = "1000") Integer size`

**Comment by marinus-suniram** [comment]
I have the same problem with the default value

**Comment by welshm** [comment]
Please see this change https://github.com/OpenAPITools/openapi-generator/pull/11181 from @cachescrubber which is addressing the issues of default params and Schema (along with other issues)

**Comment by kdebski85** [comment]
I also have the issue with defaultValue for Parameter. I created https://github.com/OpenAPITools/openapi-generator/issues/11203

**Comment by mhinders** [comment]
> > Really happy about this change. I would however be inclined to use springdocs for viewing the API definition. As it is currently the Spring controllers are generated with @org.springframework.stereotype.Controller and not @org.springframework.web.bind.annotation.RestController which means that manual configuration is required as explained here https://springdoc.org/#my-rest-controller-using-controller-annotation-is-ignored. Could the templates be changed to use RestController?
> 
> It can be updated to do that - just for clarity though, I had added this as an option for the interfaces but had left the generated API controller implementations (`apiController.mustache`) unchanged - would you be expecting both to be annotated with `@RestController`? My opinion is probably not, as the interface has all the annotations you need.

I missed that the annotation is present on the interface. That is sufficient for me.

**Comment by shivaprasadgurram** [comment]
springdoc-openapi-ui not compatible with openapi-generator-maven-plugin. I have a dependency of spring-openapi-ui with version 1.6.9. I can see generated classes are using V3 annotations but when I try to load swagger-ui.html it is not working. Not sure what is going wrong. Can any one help me? 

Need: I should be able to open swagger-ui page from browser.

**Comment by Moonergfp** [comment]
> springdoc-openapi-ui not compatible with openapi-generator-maven-plugin. I have a dependency of spring-openapi-ui with version 1.6.9. I can see generated classes are using V3 annotations but when I try to load swagger-ui.html it is not working. Not sure what is going wrong. Can any one help me?
> 
> Need: I should be able to open swagger-ui page from browser.

what would be shown when you  open swagger-ui page?

**Pull request merged**